### PR TITLE
metadata: electronicLocator

### PR DESCRIPTION
**Pull request by rerowep**
* Adds electronicLocator resource (tag 856).

Co-Authored-by: Peter Weber <peter.weber@rero.ch>

## Why are you opening this PR?

- https://tree.taiga.io/project/rero21-reroils/us/1262

## How to test?

- see in detailed view access and other access links

## Code review check list

- [ ] Commit message template compliance.
- [ ] Commit message without typos.
- [ ] File names.
- [ ] Functions names.
- [ ] Functions docstrings.
- [ ] Unnecessary commited files?

### CODE REVIEW

File: rero_ils/modules/documents/templates/rero_ils/detailed_view_documents.html [hunk 1]
```diff
@@ -123,20 +123,33 @@ <h1 class="col-sm-10 col-md-11">{{ record.title }}</h1>
...
-        {% set uris = [] %}
-        {% for resource in record.electronic_location %}
-        {% do uris.append('<a href="'+ resource.uri +'"><i class="fa fa-lock fa-fw"></i> '+ resource.uri +'</a>') %}
-        {% endfor %}
-        {{ dl(_('Online Access'), uris|join(', ')) }}
+        <!-- Other ACCESS -->
+        {% set other_accesses = record.electronicLocator|get_other_accesses %}
+        {% if other_accesses|length > 0 %}
+        <dt class="col-sm-3 offset-sm-2 offset-md-1">
+          {{ _('Other Access') }}:
```
Comment (AoNoOokami) [inline rero_ils/modules/documents/templates/rero_ils/detailed_view_documents.html]: In the mockup, it's written : Title of the fields: "Related resource" for this one

Comment (rerowep) [reply rero_ils/modules/documents/templates/rero_ils/detailed_view_documents.html]: ok

File: data/documents_big.json [hunk 2]
```diff
@@ -31042,7 +31096,7 @@
       "Inuits",
       "D\u00e9veloppement durable",
       "Tourisme culturel",
-      "Ressources naturelles",
+      "Resources naturelles",
```
Comment (AoNoOokami) [inline data/documents_big.json]: In French, it's ressources, but resource in English. I think Ressources was correct here.

File: data/documents_big.json [hunk 3]
```diff
@@ -48622,9 +48698,19 @@
     ],
     "title": "The great trade collapse : what caused it and what does it mean?",
     "notes": [
-      "Ressource en ligne consult\u00e9e le 28.02.2011"
+      "Resource en ligne consult\u00e9e le 28.02.2011"
```
Comment (AoNoOokami) [inline data/documents_big.json]: Same as previous comment

File: data/documents_big.json [hunk 4]
```diff
@@ -79238,7 +79399,7 @@
         "$ref": "https://mef.rero.ch/api/mef/12129643"
       }
     ],
-    "title": "Le grand livre des droits de l'enfant : avec le texte int\u00e9gral de la convention adopt\u00e9e par l'ONU",
+    "title": "Le grand livre des droits de l'enfant : avec le fullText de la convention adopt\u00e9e par l'ONU",
```
Comment (AoNoOokami) [inline data/documents_big.json]: fullText?

File: data/documents_big.xml [hunk 5]
```diff
@@ -12397,7 +12397,7 @@
     <subfield code="2">rdacarrier</subfield>
   </datafield>
   <datafield tag="490" ind1="1" ind2=" ">
-    <subfield code="a">Texte intégral</subfield>
+    <subfield code="a">fullText</subfield>
```
Comment (AoNoOokami) [inline data/documents_big.xml]: fulltext or Texte intégral? Present at several places in this document.

File: data/documents_small.json [hunk 6]
```diff
@@ -11379,9 +11442,19 @@
     ],
     "title": "The great trade collapse : what caused it and what does it mean?",
     "notes": [
-      "Ressource en ligne consult\u00e9e le 28.02.2011"
+      "Resource en ligne consult\u00e9e le 28.02.2011"
```
Comment (AoNoOokami) [inline data/documents_small.json]: Ressource?

File: data/documents_big.json (hunk 2 above)
Comment (rerowep) [reply data/documents_big.json]: Yes, corrected ....

File: data/documents_big.json (hunk 4 above)
Comment (rerowep) [reply data/documents_big.json]: Yes, corrected ....

File: data/documents_big.xml (hunk 5 above)
Comment (rerowep) [reply data/documents_big.xml]: In virtua it is free text, but in rero-ils we need keywords.

File: data/documents_small.json (hunk 6 above)
Comment (rerowep) [reply data/documents_small.json]: Yes, corrected ....

File: data/documents_small.xml [hunk 7]
```diff
@@ -3811,7 +3811,7 @@
     <subfield code="a">R005717371</subfield>
   </datafield>
   <datafield tag="037" ind1=" " ind2=" ">
-    <subfield code="c">Accès en texte intégral gratuit - laisdc</subfield>
+    <subfield code="c">Accès en fullText gratuit - laisdc</subfield>
```
Comment (blankoworld) [inline data/documents_small.xml]: Is that correct in french to use "fullText"? :thinking:

File: rero_ils/modules/ebooks/utils.py [hunk 8]
```diff
@@ -57,77 +57,94 @@ def add_oai_source(name, baseurl, metadataprefix='marc21',
...
-            'create document holding: {rec}'.format(
-                rec=record
             )
-        )
+    # else:
+    #     current_app.logger.warning(
+    #         'create document holding: {rec}'.format(
+    #             rec=record
+    #         )
+    #     )
```
Comment (blankoworld) [inline rero_ils/modules/ebooks/utils.py]: Why not just delete these lines?

File: data/documents_big.json (hunk 2 above)
Comment (blankoworld) [reply data/documents_big.json]: Seems that "resources naturelles" is wrong again.
Cf. https://cnrtl.fr/definition/ressource

File: data/documents_big.json [hunk 9]
```diff
@@ -12061,7 +12073,7 @@
         "source": "RERO"
       }
     ],
-    "title": "An atlas of rangeland and its main plant resources on the Qinghai-Tibet Plateau",
+    "title": "An atlas of rangeland and its main plant ressources on the Qinghai-Tibet Plateau",
```
Comment (jma) [inline data/documents_big.json]: ressources -> resources

File: rero_ils/modules/cli.py [hunk 10]
```diff
@@ -623,8 +623,8 @@ def do_worker(marc21records, results, pid_required):
         except Exception as err:
             msg = 'ERROR:\t{pid}\t{err}'.format(pid=pid, err=err.args[0])
             click.secho(msg, err=True, fg='red')
-            # import traceback
-            # traceback.print_exc()
+            if debug:
```
Comment (jma) [inline rero_ils/modules/cli.py]: is this related to this commit? Please add this, at least, in the commit message.

File: rero_ils/modules/documents/jsonschemas/documents/document-minimal-v0.0.1_src.json [hunk 11]
```diff
@@ -146,7 +146,6 @@
         ],
         "additionalProperties": false,
         "propertiesOrder": [
-          "type",
```
Comment (jma) [inline rero_ils/modules/documents/jsonschemas/documents/document-minimal-v0.0.1_src.json]: is this related to this commit?

File: rero_ils/modules/documents/jsonschemas/documents/document-minimal-v0.0.1_src.json [hunk 12]
```diff
@@ -860,29 +859,229 @@
...
+          "content",
+          "publicNote"
         ],
         "properties": {
-          "uri": {
-            "title": "Uniform Resource Identifier",
-            "description": "Uniform Resource Identifier (URI), which provides standard syntax for locating an object using existing Internet protocols.",
+          "url": {
+            "title": "url",
+            "description": "Record a unique URL here.",
```
Comment (jma) [inline rero_ils/modules/documents/jsonschemas/documents/document-minimal-v0.0.1_src.json]: Is this description comes form the PO. Seems weird. Same for other descriptions.

File: rero_ils/modules/documents/jsonschemas/documents/document-minimal-v0.0.1_src.json [hunk 13]
```diff
@@ -860,29 +859,229 @@
...
+            "placeholder": "Example: https://www.rero.ch/",
```
Comment (jma) [inline rero_ils/modules/documents/jsonschemas/documents/document-minimal-v0.0.1_src.json]: minLength is missing

File: rero_ils/modules/documents/jsonschemas/documents/document-minimal-v0.0.1_src.json [hunk 14]
```diff
@@ -860,29 +859,229 @@
...
+              ]
+            }
+          },
+          "publicNote": {
+            "title": "Uniform Resource Identifier Public Note",
+            "description": "Is displayed next to the link, as additional information",
+            "type": "array",
+            "minItems": 1,
+            "items": {
+              "type": "string",
```
Comment (jma) [inline rero_ils/modules/documents/jsonschemas/documents/document-minimal-v0.0.1_src.json]: minLength is missing

File: rero_ils/modules/documents/jsonschemas/documents/document-minimal-v0.0.1_src.json [hunk 15]
```diff
@@ -860,29 +859,229 @@
...
+              "form": {
+                "placeholder": "Example: Access only from the library"
```
Comment (jma) [inline rero_ils/modules/documents/jsonschemas/documents/document-minimal-v0.0.1_src.json]: Is it specified by the PO?

File: rero_ils/modules/documents/templates/rero_ils/detailed_view_documents.html [hunk 16]
```diff
@@ -113,8 +113,8 @@ <h1 class="col-sm-10 col-md-11">{{ record.title }}</h1>
             <li>
               {{ identifier.value }}
               <small class="badge badge-secondary text-uppercase">{{ identifier.type }}</small>
-              {% endfor %}
             </li>
+            {% endfor %}
```
Comment (jma) [inline rero_ils/modules/documents/templates/rero_ils/detailed_view_documents.html]: Can you add this fix to the commit message?

File: rero_ils/modules/documents/views.py [hunk 17]
```diff
@@ -434,3 +434,69 @@ def create_publication_statement(provision_activity):
...
+def get_cover_art(record):
+    """Get cover art.
+
+    :param record: record
+    :return: url for cover art or None
+    """
+    for electronic_locator in record.get('electronicLocator'):
+        type = electronic_locator.get('type')
+        content = electronic_locator.get('content')
+        if type and type == 'resource' and content and content == 'coverImage':
```
Comment (jma) [inline rero_ils/modules/documents/views.py]: the double check can be removed safely:
```python
if type == 'resource' and  content == 'coverImage':
```

File: rero_ils/modules/documents/views.py [hunk 18]
```diff
@@ -434,3 +434,69 @@ def create_publication_statement(provision_activity):
...
+@blueprint.app_template_filter()
+def get_accesses(record):
+    """Get electronic locator text.
+
+    :param record: record
+    :return: dictonary list of access informations
+    """
+    accesses = []
+    electronic_locators = record.get('electronicLocator', [])
+    for electronic_locator in electronic_locators:
```
Comment (jma) [inline rero_ils/modules/documents/views.py]: Looks like a filter, you can probably use this: https://docs.python.org/3/library/functions.html#filter

File: rero_ils/modules/documents/views.py [hunk 19]
```diff
@@ -434,3 +434,69 @@ def create_publication_statement(provision_activity):
...
+
+@blueprint.app_template_filter()
+def get_other_accesses(record):
+    """Length for electronic locator.
+
+    :param record: record
+    :return: dictonary list of other access informations
+    """
+    accesses = []
+    electronic_locators = record.get('electronicLocator', [])
```
Comment (jma) [inline rero_ils/modules/documents/views.py]: Same comment as previous.

File: rero_ils/modules/utils.py [hunk 20]
```diff
@@ -39,7 +39,7 @@ def strtotime(strtime):
     )
 
 
-def do_bulk_index(uuids, doc_type=None, process=False, verbose=False):
+def do_bulk_index(uuids, doc_type='rec', process=False, verbose=False):
```
Comment (jma) [inline rero_ils/modules/utils.py]: Is it related to this commit?

File: rero_ils/modules/ebooks/utils.py (hunk 8 above)
Comment (rerowep) [reply rero_ils/modules/ebooks/utils.py]: ok

File: data/documents_big.json (hunk 9 above)
Comment (rerowep) [reply data/documents_big.json]: Ok, but it is in virtua like this.

File: rero_ils/modules/cli.py (hunk 10 above)
Comment (rerowep) [reply rero_ils/modules/cli.py]: Ok

File: rero_ils/modules/documents/jsonschemas/documents/document-minimal-v0.0.1_src.json (hunk 11 above)
Comment (rerowep) [reply rero_ils/modules/documents/jsonschemas/documents/document-minimal-v0.0.1_src.json]: Yes it is for the editor.

File: rero_ils/modules/documents/jsonschemas/documents/document-minimal-v0.0.1_src.json (hunk 12 above)
Comment (rerowep) [reply rero_ils/modules/documents/jsonschemas/documents/document-minimal-v0.0.1_src.json]: Its coming from the PO

File: rero_ils/modules/documents/jsonschemas/documents/document-minimal-v0.0.1_src.json (hunk 13 above)
Comment (rerowep) [reply rero_ils/modules/documents/jsonschemas/documents/document-minimal-v0.0.1_src.json]: OK

File: rero_ils/modules/documents/templates/rero_ils/detailed_view_documents.html (hunk 16 above)
Comment (rerowep) [reply rero_ils/modules/documents/templates/rero_ils/detailed_view_documents.html]: ok

File: rero_ils/modules/documents/views.py (hunk 17 above)
Comment (rerowep) [reply rero_ils/modules/documents/views.py]: ok

File: rero_ils/modules/utils.py (hunk 20 above)
Comment (rerowep) [reply rero_ils/modules/utils.py]: It is for the tests.

File: rero_ils/modules/documents/views.py (hunk 18 above)
Comment (rerowep) [reply rero_ils/modules/documents/views.py]: Implemented with filter.

File: rero_ils/modules/documents/views.py (hunk 19 above)
Comment (rerowep) [reply rero_ils/modules/documents/views.py]: Implemented with filter.
